# Tarea 3: Sequence Labelling
**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2024)**

## Tarjeta de identificación

**Nombres:** ```Martín Reyes - Sebastián Sanhueza```

**Fecha límite de entrega 📆:** 06/06.

**Tiempo estimado de dedicación:** 4 horas


## Instrucciones
Bienvenid@s a la tercera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las últimas semanas de clases posteriores a la tarea 2, enfocado principalmente en **Word Embeddings**, **Sequence Labeling - HMM**, **Convolutional Neural Networks** y **Recurrent Neural Networks**. Si aún no has visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP.

* La tarea es en **grupo** (maximo hasta 3 personas).
* La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
* El formato de entrega es este mismo Jupyter Notebook.
* Al momento de la revisión su código será ejecutado. Por favor verifiquen que su entrega no tenga errores de compilación.
* Completar la tarjeta de identificación. Sin ella no podrá tener nota.

## Material de referencia

Diapositivas del curso 📄
    
- [Word Embeddings](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-wordvectors.pdf)
- [Sequence Labeling - HMM](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-HMM.pdf)
- [Convolutional Neural Networks](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-CNN.pdf)
- [Recurrent Neural Networks](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-RNN.pdf)

Videos del curso 📺

- Word Embeddings: [Parte 1](https://www.youtube.com/watch?v=wtwUsJMC9CA), [Parte 2](https://www.youtube.com/watch?v=XDxzQ7JU95U), [Parte 3](https://www.youtube.com/watch?v=Ikyc3DRVodk)

- Sequence Labeling - HMM: [Parte 1](https://www.youtube.com/watch?v=-ngfOZz8yK0), [Parte 2](https://www.youtube.com/watch?v=Tjgb-yQOg54), [Parte 3](https://www.youtube.com/watch?v=aaa5Qoi8Vco), [Parte 4](https://www.youtube.com/watch?v=4pKWIDkF_6Y)

- [Convolutional Neural Networks](https://www.youtube.com/watch?v=lLZW5Fn40r8)

- Recurrent Neural Networks: [Parte 1](https://www.youtube.com/watch?v=BmhjUkzz3nk), [Parte 2](https://www.youtube.com/watch?v=z43YFR1iIvk), [Parte 3](https://youtu.be/7L5JxQdwNJk)

## Objetivo

El objetivo de esta tarea es resolver una de las tareas más importantes en el área del procesamiento de lenguage natural, relacionada con la extracción de información: [Named Entity Recognition (NER)](http://www.cs.columbia.edu/~mcollins/cs4705-spring2019/slides/tagging.pdf).

En particular, deberán crear distintos modelos que apunten a resolver la tarea de NER en Español. Para esto, les entregaremos un dataset real perteneciente a la lista de espera NO GES en Chile. Es importante destacar que existe una falta de trabajos realizados en el área de NER en Español y aún más en el contexto clínico, por ende puede ser considerado como una tarea bien desafiante y quizás les interesa trabajar en el área más adelante en sus carreras.

En este notebook les entregaremos un baseline como referencia de los resultados que esperamos puedan obtener. Recuerden que el no superar a los baselines en alguna de las tres métricas conlleva un descuento de 0.5 puntos hasta 1.5 puntos.

Como hemos estado viendo redes neuronales tanto en cátedras, tareas y auxiliares (o próximamente lo harán), esperamos que (por lo menos) utilicen Redes Neuronales Recurrentes (RNN) para resolverla.

Nuevamente, hay total libertad para utilizar el software y los modelos que deseen, siempre y cuando estos no traigan los modelos ya implementados (de todas maneras, como es un corpus nuevo, es difícil que haya algún modelo ya implementado con éstas entidades).

## Explicación de NER

En esta tarea van a resolver **NER**, comúnmente abordada como un problema de Sequence Labeling.

**¿Qué es Sequence Labeling?**

En breves palabras, dada una secuencia de tokens (oración) sequence labeling tiene por objetivo asignar una etiqueta a cada token de dicha secuencia. En pocas palabras, dada una lista de tokens esperamos encontrar la mejor secuencia de etiquetas asociadas a esa lista. Ahora veamos de qué se trata este problema.

**Named Entity Recognition (NER)**

NER es un ejemplo de un problema de Sequence Labeling. Pero antes de definir formalmente esta tarea, es necesario definir algunos conceptos claves para poder entenderla de la mejor manera:

- *Token*: Un token es una secuencia de caracteres, puede ser una palabra, un número o un símbolo.

- *Entidad*: No es más que un trozo de texto (uno o más tokens) asociado a una categoría predefinida. Originalmente se solían utilizar categorías como nombres de personas, organizaciones, ubicaciones, pero actualmente se ha extendido a diferentes dominios.

- *Límites de una entidad*: Son los índices de los tokens de inicio y fín dentro de una entidad.

- *Tipo de entidad*: Es la categoría predefinida asociada a la entidad.

Dicho esto, definimos formalmente una entidad como una tupla: $(s, e, t)$, donde $s, t$ son los límites de la entidad (índices de los tokens de inicio y fin, respectivamente) y corresponde al tipo de entidad o categoría. Ya veremos más ejemplos luego de describir el Dataset.

**Corpus de la Lista de espera**

Trabajaran con un conjunto de datos reales correspondiente a interconsultas de la lista de espera NO GES en Chile. Si quieren saber más sobre cómo fueron generados los datos pueden revisar el paper publicado hace unos meses atrás en el workshop de EMNLP, una de las conferencias más importantes de NLP: [https://www.aclweb.org/anthology/2020.clinicalnlp-1.32/](https://www.aclweb.org/anthology/2020.clinicalnlp-1.32/).

Este corpus Chileno está constituido originalmente por 7 tipos de entidades pero por simplicidad en esta tarea trabajarán con las siguientes:

- **Disease**
- **Body_Part**
- **Medication**
- **Procedures**
- **Family_Member**

Si quieren obtener más información sobre estas entidades pueden consultar la [guía de anotación](https://plncmm.github.io/annodoc/). Además, mencionar que este corpus está restringido bajo una licencia que permite solamente su uso académico, así que no puede ser compartido más allá de este curso o sin permisos por parte de los autores en caso que quieran utilizarlo fuera. Si este último es el caso entonces pueden escribir directamente al correo: pln@cmm.uchile.cl. Al aceptar los términos y condiciones de la tarea están de acuerdo con los puntos descritos anteriormente.


**Formato ConLL**

Los archivos que serán entregados a ustedes vienen en un formato estándar utilizado en NER, llamado ConLL. No es más que un archivo de texto, que cumple las siguientes propiedades.

- Un salto de linea corresponde a la separación entre oraciones. Esto es importante ya que al entrenar una red neuronal ustedes pasaran una lista de oraciones como input, más conocidos como batches.

- La primera columna del archivo contiene todos los tokens de la partición.

- La segunda columna del archivo contiene el tipo de entidad asociado al token de la primera columna.

- Los tipos de entidades siguen un formato clásico en NER denominado *IOB2*. Si un tipo de entidad comienza con el prefijo "B-" (Beginning) significa que es el token de inicio de una entidad, si comienza con "I-" (Inside) es un token distinto al de inicio y si un token está asociado a la categoría O (Outside) significa que no pertenece a ninguna entidad.

Aquí va un ejemplo:

```
PACIENTE O
PRESENTA O
FRACTURA B-Disease
CORONARIA I-Disease
COMPLICADA I-Disease
EN O
PIE B-Body_Part
IZQUIERDO I-Body_Part
. O
SE O
REALIZA O
INSTRUMENTACION B-Procedure
INTRACONDUCTO I-Procedure
. O
```

Según nuestra definición tenemos las siguientes tres entidades (enumerando desde 0):

- $(2, 4, Disease)$
- $(6, 7, Body Part)$
- $(11, 12, Procedure)$

Repasen un par de veces todos estos conceptos antes de pasar a la siguiente sección del notebook.
Es importante entender bien este formato ya que al medir el rendimiento de sus modelos, consideraremos una **métrica estricta**. Esta métrica se llama así ya que considera correcta una predicción de su modelo, sólo si al compararlo con las entidades reales **coinciden tanto los límites de la entidad como el tipo.**

Para ejemplificar, tomando el caso anterior, si el modelo es capaz de encontrar la siguiente entidad: $(2, 3, Disease)$, entonces se considera incorrecto ya que pudo predecir dos de los tres tokens de dicha enfermedad. Es decir, buscamos una métrica que sea alta a nivel de entidad y no a nivel de token.

Antes de pasar a explicar las reglas, se recomienda visitar los siguientes links para entender bien el baseline de la tarea:

-  [Recurrent Neural Networks](slides/NLP-RNN.pdf): [Parte 1](https://youtu.be/BmhjUkzz3nk), [Parte 2](https://youtu.be/z43YFR1iIvk), [Parte 3](https://youtu.be/7L5JxQdwNJk)


Recuerden que todo el material se encuentra disponible en el [github del curso](https://github.com/dccuchile/CC6205).

## Baseline

En este punto esperamos que tengan conocimiento sobre redes neuronales y en particular redes neuronales recurrentes (RNN), si no siempre pueden escribirnos por el canal de Discord para aclarar dudas. La RNN del baseline adjunto a este notebook está programado en la librería [`pytorch`](https://pytorch.org/) pero ustedes pueden utilizar keras, tensorflow si así lo desean. El código contiene lo siguiente:

- La carga de los datasets, creación de batches de texto y padding (esto es importante ya que si utilizan redes neuronales tienen que tener el mismo largo los inputs).

- La implementación básica de una red `LSTM` simple de solo un nivel y sin bi-direccionalidad.

- La construcción del formato del output requerido para que lo puedan probar en la tarea en codalab.

### **Sugerencias**
Se espera que ustedes puedan experimentar con el baseline utilizando (pero no limitándose) estas sugerencias:

*   Probar la técnica de early stopping.
*   Variar la cantidad de parámetros de la capa de embeddings.
*   Variar la cantidad de capas RNN.
*   Variar la cantidad de parámetros de las capas de RNN.
*   Inicializar la capa de embeddings con modelos pre-entrenados. (word2vec, glove, conceptnet, etc...). [Embeddings en español aquí](https://github.com/dccuchile/spanish-word-embeddings). También aquí pueden encontrar unos embeddings clínicos en Español: [https://zenodo.org/record/3924799](https://zenodo.org/record/3924799)
*   Variar la cantidad de épocas de entrenamiento.
*   Variar el optimizador, learning rate, batch size, etc.
*   Probar bi-direccionalidad.
*   Incluir dropout.
*   Probar modelos de tipo GRU.
*   Probar usando capas de atención.
*   Probar Embedding Contextuales (les puede ser de utilidad [flair](https://github.com/flairNLP/flair))
*   Probar modelos de transformers en español usando [Huggingface](https://github.com/huggingface/transformers) o el framework Flair.

## Experimento base

El código que les entregaremos servirá de baseline para luego implementar mejores modelos.
En general, el código asociado a la carga de los datos, las funciones de entrenamiento, de evaluación y la predicción de los datos de la tarea no deberían cambiar.
Solo deben preocuparse de cambiar la arquitectura del modelo, sus hiperparámetros y reportar, lo cual lo pueden hacer en las subsecciones *modelos*.

###  **Carga de datos y Preprocesamiento**

Para cargar los datos y preprocesarlos usaremos la librería [`torchtext`](https://github.com/pytorch/text). Tener cuidado ya que hace algunos meses esta librería tuvo cambios radicales, quedando las funcionalidades pasadas depreciadas de la librería ```legacy```. Esto ya que si quieren usar más funciones de la librería entonces vean los cambios en la documentación debe usar la versión antigua con python 3.8

El proceso será el siguiente:

1. Descargar los datos desde github y examinarlos.
2. Cargar los datasets con la clase ```TaggingDataset``` de más abajo.
3. Crear el vocabulario.

In [ ]:
%%capture
!pip install -U torchtext

In [ ]:
import torch
import torchtext
from torchtext import data, datasets

# Garantizar reproducibilidad de los experimentos
SEED = 1234
torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

/opt/conda/lib/python3.10/site-packages/torchtext/data/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)
/opt/conda/lib/python3.10/site-packages/torchtext/datasets/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)


#### **Obtener datos**

Descargamos los datos de entrenamiento, validación y prueba en nuestro directorio de trabajo

In [ ]:
%%capture

!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/train.txt -nc
!wget https://github.com/dccuchile/CC6205/releases/download/v1.0/dev.txt -nc

In [ ]:
# NUEVO DATALOADER Y OTRAS COSAS NECESARIAS
from collections import Counter, OrderedDict

import torch

from torch.utils.data import DataLoader, Dataset
from torch.nn.utils.rnn import pad_sequence
from torchtext.vocab import vocab

class TaggingDataset(Dataset):
    def __init__(self, paths, lower=False, separator=" ", encoding="utf-8"):

        data = []
        for path in paths:
          with open(path, 'r', encoding=encoding) as file:
            text, tag = [], []
            for line in file:
                line = line.strip()
                if line == "":
                    data.append(dict({'text':text, 'nertags':tag}))
                    text, tag = [], []
                else:
                    line_content = line.split(separator) # .rstrip('\n')
                    if lower:
                      text.append(line_content[0].lower())
                    else:
                      text.append(line_content[0])
                    tag.append(line_content[1])
          data.append(dict({'text':text, 'nertags':tag}))

        self.data = data

    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        item = self.data[index]
        text = item["text"]
        nertags = item["nertags"]
        return nertags, text

def fit_vocab(data_iter):

  def update_counter(counter_obj):
    sorted_by_freq_tuples = sorted(counter_obj.items(),
                                  key=lambda x: x[1],
                                  reverse=True)
    ordered_dict = OrderedDict(sorted_by_freq_tuples)
    return ordered_dict

  counter_1 = Counter()
  counter_2 = Counter()
  for _nertags, _text in data_iter:
    counter_1.update(_text)
    counter_2.update(_nertags)

  od1 = update_counter(counter_1)
  od2 = update_counter(counter_2)

  v1 = vocab(od1, specials=['<PAD>', '<unk>'])
  v1.set_default_index(v1["<unk>"])
  v2 = vocab(od2, specials=['<PAD>'])

  text_pipeline = lambda x: v1(x)
  nertags_pipeline = lambda x: v2(x)

  return text_pipeline, nertags_pipeline, v1, v2

def collate_batch(batch, nertags_pipeline, text_pipeline, device):
  nertags_list, text_list = [], []
  for _nertags, _text in batch:
    processed_nertags = torch.tensor(nertags_pipeline(_nertags),
                                     dtype=torch.int64)
    nertags_list.append(processed_nertags)
    processed_text = torch.tensor(text_pipeline(_text),
                                  dtype=torch.int64)
    text_list.append(processed_text)
  nertags_list = pad_sequence(nertags_list, batch_first=True).T
  text_list = pad_sequence(text_list, batch_first=True).T
  return nertags_list.to(device), text_list.to(device)

/opt/conda/lib/python3.10/site-packages/torchtext/vocab/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)
/opt/conda/lib/python3.10/site-packages/torchtext/utils.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)


In [ ]:
data_iter = TaggingDataset(["train.txt", "dev.txt"])

data_length = len(data_iter)

train_length = int(data_length * 0.8)
dev_length = int(data_length * 0.1)
test_length = data_length - train_length - dev_length

train_iter, dev_iter, test_iter = torch.utils.data.random_split(
    data_iter,
     (train_length, dev_length, test_length),
    torch.Generator().manual_seed(42))

text_pipeline, nertags_pipeline, TEXT, NER_TAGS = fit_vocab(train_iter)

In [ ]:
len(train_iter), len(dev_iter), len(test_iter)

(7132, 891, 893)

In [ ]:
# seteamos algunos valores de interes
UNK_IDX = TEXT.vocab.get_stoi()['<unk>']
PAD_IDX = TEXT.vocab.get_stoi()['<PAD>']

PAD_TAG_IDX = NER_TAGS.get_stoi()['<PAD>']
O_TAG_IDX = NER_TAGS.vocab.get_stoi()['O']

In [ ]:
BATCH_SIZE = 22

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using', device)

fix_collocate_batch = lambda x: collate_batch(x, nertags_pipeline, text_pipeline, device)

dataloader_train = DataLoader(
    train_iter, batch_size=BATCH_SIZE, shuffle=False, collate_fn=fix_collocate_batch
)
dataloader_dev = DataLoader(
    dev_iter, batch_size=BATCH_SIZE, shuffle=False, collate_fn=fix_collocate_batch
)
dataloader_test = DataLoader(
    test_iter, batch_size=BATCH_SIZE, shuffle=False, collate_fn=fix_collocate_batch
)

Using cuda


In [ ]:
example = next(iter(dataloader_train))
ner_tags_example = example[0]
text_example = example[1]

# revisamos el primer ejemplo
[NER_TAGS.vocab.get_itos()[j] for j in ner_tags_example[:, 1]], [TEXT.vocab.get_itos()[j] for j in text_example[:, 1]]

(['O',
  'O',
  'O',
  'O',
  'B-Disease',
  'I-Disease',
  'I-Disease',
  'I-Disease',
  'O',
  'O',
  'O',
  'O',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>'],
 ['SOLICITO',
  'EVALUACION',
  'Y',
  'TTO',
  'SUBLUXACION',
  'ATM',
  'IZQ',
  '.',
  ',',
  'DOLOR',
  'APERTURA',
  '.',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>',
  '<PAD>'])

In [ ]:
%%capture
!pip install seqeval

In [ ]:
# Definimos las métricas

from seqeval.metrics import f1_score, precision_score, recall_score

def calculate_metrics(preds, y_true, pad_idx=PAD_TAG_IDX, o_idx=O_TAG_IDX):
    """
    Calcula precision, recall y f1 de cada batch.
    """

    # Obtener el indice de la clase con probabilidad mayor. (clases)
    y_pred = preds.argmax(dim=1, keepdim=True)

    # filtramos <pad> para calcular los scores.
    mask = [(y_true != pad_idx)]
    y_pred = y_pred[mask]
    y_true = y_true[mask]

    # traemos a la cpu
    y_pred = y_pred.view(-1).to('cpu').numpy()
    y_true = y_true.to('cpu').numpy()
    y_pred = [[NER_TAGS.vocab.get_itos()[v] for v in y_pred]]
    y_true = [[NER_TAGS.vocab.get_itos()[v] for v in y_true]]

    # calcular scores
    f1 = f1_score(y_true, y_pred, mode='strict')
    precision = precision_score(y_true, y_pred, mode='strict')
    recall = recall_score(y_true, y_pred, mode='strict')

    return precision, recall, f1

### **Modelo Baseline**

Teniendo ya cargado los datos, toca definir nuestro modelo. Este baseline tendrá una capa de embedding, unas cuantas LSTM y una capa de salida y usará dropout en el entrenamiento.

Este constará de los siguientes pasos:

1. Definir la clase que contendrá la red.
2. Definir los hiperparámetros e inicializar la red.
3. Definir el número de épocas de entrenamiento
4. Definir la función de loss.



Recomendamos que para experimentar, encapsules los modelos en una sola variable y luego la fijes en model para entrenarla

In [ ]:
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim


# Definir la red
class NER_RNN(nn.Module):
    def __init__(self,
                 input_dim,
                 embedding_dim,
                 hidden_dim,
                 output_dim,
                 n_layers,
                 bidirectional,
                 dropout,
                 pad_idx):

        super().__init__()

        # Capa de embedding
        self.embedding = nn.Embedding(input_dim,
                                      embedding_dim,
                                      padding_idx=pad_idx,
                                      )

        # Capa LSTM
        self.lstm = nn.LSTM(embedding_dim,
                           hidden_dim,
                           num_layers=n_layers,
                           bidirectional=bidirectional,
                           dropout = dropout if n_layers > 1 else 0)

        # Capa de salida
        self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim,
                            output_dim)

        # Dropout
        self.dropout = nn.Dropout(dropout)

    def forward(self, text):

        #text = [sent len, batch size]

        # Convertir lo enviado a embedding
        embedded = self.dropout(self.embedding(text))

        outputs, (hidden, cell) = self.lstm(embedded)
        #embedded = [sent len, batch size, emb dim]

        # Pasar los embeddings por la rnn (LSTM)

        #output = [sent len, batch size, hid dim * n directions]
        #hidden/cell = [n layers * n directions, batch size, hid dim]

        # Predecir usando la capa de salida.
        predictions = self.fc(self.dropout(outputs))
        #predictions = [sent len, batch size, output dim]

        return predictions

Hiperparámetros de la red

In [ ]:
# tamaño del vocabulario. recuerden que la entrada son vectores bag of word(one-hot).
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 300  # dimensión de los embeddings.
HIDDEN_DIM = 256  # dimensión de la capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab)  # número de clases

N_LAYERS = 1  # número de capas.
DROPOUT = 0.0
BIDIRECTIONAL = False

# Creamos nuestro modelo.
baseline_model = NER_RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM,
                         N_LAYERS, BIDIRECTIONAL, DROPOUT, PAD_IDX)

baseline_model_name = 'baseline'  # nombre que tendrá el modelo guardado...

baseline_n_epochs = 10

Definimos la función de loss

In [ ]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.get_stoi()['<PAD>']
baseline_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX)

### **Entrenamos y evaluamos**


**Importante** : Fijen el modelo, el número de épocas de entrenamiento, la loss y el optimizador que usarán para entrenar y evaluar en las siguientes variables!!!

In [ ]:
model = baseline_model
model_name = baseline_model_name
criterion = baseline_criterion
n_epochs = baseline_n_epochs

#### **Inicializamos la red**

Iniciamos los pesos de la red de forma aleatoria (Usando una distribución normal).

In [ ]:
def init_weights(m):
    # Inicializamos los pesos como aleatorios
    for name, param in m.named_parameters():
        nn.init.normal_(param.data, mean=0, std=0.1)

    # Seteamos como 0 los embeddings de UNK y PAD.
    model.embedding.weight.data[UNK_IDX] = torch.zeros(EMBEDDING_DIM)
    model.embedding.weight.data[PAD_IDX] = torch.zeros(EMBEDDING_DIM)

model.apply(init_weights)

NER_RNN(
  (embedding): Embedding(16496, 300, padding_idx=0)
  (lstm): LSTM(300, 256)
  (fc): Linear(in_features=256, out_features=12, bias=True)
  (dropout): Dropout(p=0.0, inplace=False)
)

In [ ]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f'El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

El modelo actual tiene 5,523,276 parámetros entrenables.


Notar que definimos los embeddings que representan a \<unk\> y \<pad\>  como [0, 0, ..., 0]

Definimos el optimizador

In [ ]:
# Optimizador
optimizer = optim.Adam(model.parameters()) # SGD, AdamW

Enviamos el modelo a cuda

In [ ]:
# Enviamos el modelo y la loss a cuda (en el caso en que esté disponible)
model = model.to(device)
criterion = criterion.to(device)

#### **Definimos el entrenamiento de la red**

Algunos conceptos previos:

- `epoch` : una pasada de entrenamiento completa de una dataset.
- `batch`: una fracción de la época. Se utilizan para entrenar mas rápidamente la red. (mas eficiente pasar n datos que uno en cada ejecución del backpropagation)

Esta función está encargada de entrenar la red en una época. Para esto, por cada batch de la época actual, predice los tags del texto, calcula su loss y luego hace backpropagation para actualizar los pesos de la red.

Observación: En algunos comentarios aparecerá el tamaño de los tensores entre corchetes

In [ ]:
def train(model, iterator, optimizer, criterion):
    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.train()

    # Por cada batch del iterador de la época:
    for tags, text in iterator:
        # Reiniciamos los gradientes calculados en la iteración anterior
        optimizer.zero_grad()

        #text = [sent len, batch size]

        # Predecimos los tags del texto del batch.
        predictions = model(text.to(device))

        #predictions = [sent len, batch size, output dim]
        #tags = [sent len, batch size]

        # Reordenamos los datos para calcular la loss
        predictions = predictions.view(-1, predictions.shape[-1])
        #ipdb.set_trace()
        tags = torch.reshape(tags, (-1,)).to(device)

        #predictions = [sent len * batch size, output dim]

        # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
        loss = criterion(predictions, tags)

        # Calculamos el accuracy
        precision, recall, f1 = calculate_metrics(predictions, tags)

        # Calculamos los gradientes
        loss.backward()

        # Actualizamos los parámetros de la red
        optimizer.step()

        # Actualizamos el loss y las métricas
        epoch_loss += loss.item()
        epoch_precision += precision
        epoch_recall += recall
        epoch_f1 += f1

    return ( epoch_loss / len(iterator), epoch_precision / len(iterator),
              epoch_recall / len(iterator), epoch_f1 / len(iterator) )

#### **Definimos la función de evaluación**

Evalua el rendimiento actual de la red usando los datos de validación.

Por cada batch de estos datos, calcula y reporta el loss y las métricas asociadas al conjunto de validación.
Ya que las métricas son calculadas por cada batch, estas son retornadas promediadas por el número de batches entregados. (ver linea del return)

In [ ]:
def evaluate(model, iterator, criterion):

    epoch_loss = 0
    epoch_precision = 0
    epoch_recall = 0
    epoch_f1 = 0

    model.eval()

    # Indicamos que ahora no guardaremos los gradientes
    with torch.no_grad():
        # Por cada batch
        for tags, text in iterator:
            # Predecimos
            predictions = model(text.to(device))

            predictions = predictions.view(-1, predictions.shape[-1])
            tags = torch.reshape(tags, (-1,)).to(device)

            # Calculamos el Cross Entropy de las predicciones con respecto a las etiquetas reales
            loss = criterion(predictions, tags)

            # Calculamos las métricas
            precision, recall, f1 = calculate_metrics(predictions, tags)

            # Actualizamos el loss y las métricas
            epoch_loss += loss.item()
            epoch_precision += precision
            epoch_recall += recall
            epoch_f1 += f1

    return epoch_loss / len(iterator), epoch_precision / len(
        iterator), epoch_recall / len(iterator), epoch_f1 / len(iterator)

In [ ]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

#### **Entrenamiento de la red**

En este cuadro de código ejecutaremos el entrenamiento de la red.
Para esto, primero definiremos el número de épocas y luego por cada época, ejecutaremos `train` y `evaluate`.

**Importante: Reiniciar los pesos del modelo**

Si ejecutas nuevamente esta celda, se seguira entrenando el mismo modelo una y otra vez.
Para reiniciar el modelo se debe ejecutar nuevamente la celda que contiene la función `init_weights`.

In [ ]:
best_valid_loss = float('inf')

for epoch in range(n_epochs):

    start_time = time.time()

    # Recuerdo: dataloader_train y valid_iterator contienen el dataset dividido en batches

    # Entrenar
    train_loss, train_precision, train_recall, train_f1 = train(
        model, dataloader_train, optimizer, criterion)

    # Evaluar (valid = validación)
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model, dataloader_dev, criterion)

    end_time = time.time()

    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Si obtuvimos mejores resultados, guardamos este modelo en el almacenamiento (para poder cargarlo luego)
    # Si detienen el entrenamiento prematuramente, pueden cargar el modelo en el siguiente recuadro de código
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), '{}.pt'.format(model_name))
    # Si ya no mejoramos el loss de validación, terminamos de entrenar

    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

/opt/conda/lib/python3.10/site-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: <PAD> seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))


Epoch: 01 | Epoch Time: 0m 7s
	Train Loss: 0.716 | Train f1: 0.48 | Train precision: 0.61 | Train recall: 0.41
	 Val. Loss: 0.435 |  Val. f1: 0.70 |  Val. precision: 0.75 | Val. recall: 0.66
Epoch: 02 | Epoch Time: 0m 6s
	Train Loss: 0.314 | Train f1: 0.77 | Train precision: 0.80 | Train recall: 0.74
	 Val. Loss: 0.373 |  Val. f1: 0.73 |  Val. precision: 0.76 | Val. recall: 0.71
Epoch: 03 | Epoch Time: 0m 6s
	Train Loss: 0.193 | Train f1: 0.86 | Train precision: 0.87 | Train recall: 0.85
	 Val. Loss: 0.392 |  Val. f1: 0.74 |  Val. precision: 0.78 | Val. recall: 0.70
Epoch: 04 | Epoch Time: 0m 6s
	Train Loss: 0.136 | Train f1: 0.90 | Train precision: 0.90 | Train recall: 0.90
	 Val. Loss: 0.427 |  Val. f1: 0.74 |  Val. precision: 0.78 | Val. recall: 0.70
Epoch: 05 | Epoch Time: 0m 6s
	Train Loss: 0.102 | Train f1: 0.93 | Train precision: 0.93 | Train recall: 0.93
	 Val. Loss: 0.449 |  Val. f1: 0.74 |  Val. precision: 0.79 | Val. recall: 0.70
Epoch: 06 | Epoch Time: 0m 6s
	Train Loss: 0.

**Importante**: Recuerden que el último modelo entrenado no es el mejor (probablemente esté *overfitteado*), si no el que guardamos con la menor loss del conjunto de validación. Este problema lo pueden solucionar con *early stopping*.
Para cargar el mejor modelo entrenado, ejecuten la siguiente celda.

In [ ]:
# cargar el mejor modelo entrenado.
model.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [ ]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

#### **Evaluamos el set de validación con el modelo final**

Estos son los resultados de predecir el dataset de evaluación con el *mejor* modelo entrenado.

In [ ]:
valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
    model, dataloader_dev, criterion)

print(
    f'Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} | Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
)

Val. Loss: 0.373 |  Val. f1: 0.73 | Val. precision: 0.76 | Val. recall: 0.71


#### **Evaluamos el set de prueba con el modelo final**

Estos son los resultados de predecir el dataset de prueba con el *mejor* modelo entrenado.

In [ ]:
prueba_loss, prueba_precision, prueba_recall, prueba_f1 = evaluate(
    model, dataloader_test, criterion)

print(
    f'Test. Loss: {prueba_loss:.3f} |  Test. f1: {prueba_f1:.2f} | Test. precision: {prueba_precision:.2f} | Test. recall: {prueba_recall:.2f}'
)

Test. Loss: 0.393 |  Test. f1: 0.73 | Test. precision: 0.76 | Test. recall: 0.71


## Experimentos

En esta sección deberan explicar, implementar y evaluar **TRES EXPERIMENTO DIFERENTES**. Puede revisar la [lista de sugerencias](#scrollTo=TfMlGTCgIq8o) al inicio del enunciado para sacar ideas de experimentos. Se espera que los experimentos sean relevantes y desafiantes.

Cada experimento debe contener las siguientes cuatro subsecciones:

- *Explicación del experimento*: debe motivar y explicar en que consiste su experimento. La explicación debe fundamentar porque es un experimento relevante y desafiantes, ademas de aclarar que estudiaran.

- *Implementacion del experimento*: debe implementar su experimento con codigo debidamente comentado, con tal de que sea legible.

- *Evaluar el experimento*: debe evaluar con las metricas propuestas para generar los resultados obtenidos con sus experimentos sobre el conjunto de prueba.

- *Analizar el experimento*: debe hacer un breve analisis de los resultados obtenidos.

In [ ]:
# Importar librerias
import numpy as np
from gensim.models import KeyedVectors

### **Experimento 1**

#### Explicación

Cambios con respecto al modelo baseline:

- Aplicación de la capa de embeddings preentrenados de "Chilean Waiting List".
- Reemplazo de la capa LSTM por una capa GRU.
- Inclusión de una capa de dropout con una tasa de 0.6.
- Configuración de la capa GRU como bidireccional.

En este experimento, se integraron embeddings pre-entrenados de "Chilean Waiting List", lo que no solo enriquece la calidad de las representaciones de las palabras, sino que también acelera el proceso de entrenamiento al proporcionar un punto de partida sólido. Luego, se reemplazó la capa LSTM por una capa GRU, que es menos compleja computacionalmente y puede manejar mejor la desaparición del gradiente. Además, se añadió una capa de dropout con una tasa de 0.6 para prevenir el sobreajuste, mejorando así la generalización del modelo. Finalmente, la configuración bidireccional de la capa GRU permite capturar dependencias contextuales tanto del pasado como del futuro en la secuencia de texto, enriqueciendo la capacidad del modelo para reconocer entidades nombradas con mayor precisión.

Este experimento es relevante para evaluar el tradeoff entre el rendimiento de un modelo GRU bidireccional y un modelo LSTM unidireccional, manteniendo una cantidad de parámetros similar.

#### Implementación

In [ ]:
class NER_RNN_EMB_GRU(nn.Module):
      """
      Modelo de red neuronal para tarea de Name Entity Recognition (NER)
      utilizando capas GRU con embeddings pre-entrenados.

      Parameters
      ----------
      embedding_dim : int
              Dimensión de los embeddings.
      hidden_dim : int
              Dimensión de las capas ocultas de GRU.
      output_dim : int
              Dimensión de la salida (número de etiquetas de NER).
      n_layers : int
              Número de capas de GRU.
      bidirectional : bool
              Si la GRU es bidireccional.
      dropout : float
              Tasa de dropout.
      pad_idx : int
              Índice de padding en el vocabulario.
      pretrained_embeddings : Tensor
              Embeddings pre-entrenados a utilizar.
      """
      def __init__(self,
                  embedding_dim,
                  hidden_dim,
                  output_dim,
                  n_layers,
                  bidirectional,
                  dropout,
                  pad_idx,
                  pretrained_embeddings
                  ):

          super(NER_RNN_EMB_GRU, self).__init__()

          # Capa de embeddings utilizando embeddings pre-entrenados
          self.embedding = nn.Embedding.from_pretrained(pretrained_embeddings,
                                                        freeze=False,
                                                        padding_idx=pad_idx
                                                        )
          # Dimensión de los embeddings
          self.embedding_dim = embedding_dim

          # Capa GRU
          self.gru = nn.GRU(self.embedding_dim,
                            hidden_dim,
                            num_layers=n_layers,
                            bidirectional=bidirectional,
                            dropout=dropout if n_layers > 1 else 0
                            )

          # Capa de salida
          self.fc = nn.Linear(hidden_dim * 2 if bidirectional else hidden_dim,
                              output_dim
                              )

          # Capa de Dropout
          self.dropout = nn.Dropout(dropout)

      def forward(self, text):
          """
          Define el paso hacia adelante de la red.

          Parameters
          ----------
          text : Tensor
              Tensor de entrada de tamaño [sent len, batch size].

          Returns
          -------
          Tensor :
              Predicciones de la red de tamaño [sent len, batch size, output dim].
          """
          # Aplicar embeddings y dropout
          embedded = self.dropout(self.embedding(text)) # embedded = [sent len, batch size, emb dim]

          # Pasar los embeddings por la GRU
          outputs, h_n = self.gru(embedded) # outputs = [sent len, batch size, hid dim * n directions]

          # Pasar las salidas por la capa de dropout y la capa lineal
          predictions = self.fc(self.dropout(outputs)) # predictions = [sent len, batch size, output dim]

          return predictions

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Ruta al archivo de embeddings pre-entrenados
pretrained_embedding_path = '/content/drive/MyDrive/cwlce.vec'

# Cargar embeddings pre-entrenados en formato word2vec
word_vectors = KeyedVectors.load_word2vec_format(pretrained_embedding_path, binary=False)

# Obtener vocabulario
vocab = TEXT.vocab

count=[]
# Asignar embeddings pre-entrenados a los tokens del vocabulario
for token, idx in vocab.get_stoi().items():
    # Verificar si el token está en los embeddings pre-entrenados
    if token.lower() in word_vectors.key_to_index:
        count.append(idx)
print(f"Cantidad de tokens coincidentes: {len(count)}")

Cantidad de tokens coincidentes: 12947


In [ ]:
# Dimensión de los embeddings pre-entrenados
embedding_dim = word_vectors.vector_size

# Inicializar la matriz de embeddings con ceros
embedding_matrix = torch.zeros(len(TEXT), embedding_dim)

# Asignar embeddings pre-entrenados a los tokens del vocabulario
for token, idx in TEXT.get_stoi().items():
    # Verificar si el token está en los embeddings pre-entrenados
    if token.lower() in word_vectors.key_to_index:
        embedding_matrix[idx] = torch.tensor(np.copy(word_vectors[token.lower()]), dtype=torch.float)
    elif token == '<PAD>':
        # Si el token es '<PAD>', asignar un vector de ceros
        embedding_matrix[idx] = torch.zeros(embedding_dim)
    else:
        # Si el token no está en los embeddings pre-entrenados, asignar un vector aleatorio
        embedding_matrix[idx] = torch.randn(embedding_dim)

In [ ]:
# INPUT_DIM = len(TEXT.vocab)
PRETRAINED_EMBEDDINGS = embedding_matrix # Utilizamos la matriz de embeddings preentrenados creada anteriormente
EMBEDDING_DIM = embedding_dim # Dimensión de los embeddings
HIDDEN_DIM = 256 # Dimensión de las capas GRU
OUTPUT_DIM = len(NER_TAGS.vocab) # Número de clases

N_LAYERS = 1  # Número de capas GRU
DROPOUT = 0.6 # Tasa de Dropout
BIDIRECTIONAL = True # Indicador de si la GRU es bidireccional

# Crear una instancia del modelo NER_RNN_EMB_GRU
model_1 = NER_RNN_EMB_GRU(embedding_dim=EMBEDDING_DIM,
                          hidden_dim=HIDDEN_DIM,
                          output_dim=OUTPUT_DIM,
                          n_layers=N_LAYERS,
                          bidirectional=BIDIRECTIONAL,
                          dropout=DROPOUT,
                          pad_idx=PAD_IDX,
                          pretrained_embeddings=PRETRAINED_EMBEDDINGS).to(device)

# Nombre que tendrá el modelo guardado
model_1_name = 'experimento_1'
# Número de épocas para entrenar el modelo
model_1_n_epochs = 10

In [ ]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.get_stoi()['<PAD>'] # Definir el índice de padding para las etiquetas
model_1_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX) # Definir la función de pérdida ignorando el índice de padding

In [ ]:
# Asignar el modelo, el nombre del modelo, la función de pérdida y el número de épocas de entrenamiento
model = model_1
model_name = model_1_name
model_criterion = model_1_criterion
model_n_epochs = model_1_n_epochs

# Imprimir el número de parámetros entrenables en el modelo actual
print(f'El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

El modelo actual tiene 5,812,044 parámetros entrenables.


In [ ]:
# Aplicar la inicialización de pesos al modelo
model.apply(init_weights)

NER_RNN_EMB_GRU(
  (embedding): Embedding(16496, 300, padding_idx=0)
  (gru): GRU(300, 256, bidirectional=True)
  (fc): Linear(in_features=512, out_features=12, bias=True)
  (dropout): Dropout(p=0.6, inplace=False)
)

In [ ]:
# Optimizador
model_optimizer = optim.Adam(model.parameters()) # SGD, AdamW

In [ ]:
# Enviar el modelo y la función de pérdida a la GPU (si está disponible)
model = model.to(device)
model_criterion = model_criterion.to(device)

In [ ]:
best_valid_loss = float('inf') # Inicializar la mejor pérdida de validación como infinita

# Bucle de entrenamiento y validación para cada época
for epoch in range(model_n_epochs):

    start_time = time.time()

    # Entrenar el modelo con el conjunto de entrenamiento
    train_loss, train_precision, train_recall, train_f1 = train(
        model, dataloader_train, model_optimizer, model_criterion)

    # Evaluar el modelo en el conjunto de validación
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model, dataloader_dev, model_criterion)

    end_time = time.time()
    # Calcular el tiempo de la época (minutos y segundos)
    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Guardar el modelo si la pérdida de validación es la mejor obtenida hasta ahora
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), '{}.pt'.format(model_name))

    # Imprimir resultados de la época
    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

/opt/conda/lib/python3.10/site-packages/seqeval/metrics/sequence_labeling.py:171: UserWarning: <PAD> seems not to be NE tag.
  warnings.warn('{} seems not to be NE tag.'.format(chunk))
/opt/conda/lib/python3.10/site-packages/seqeval/metrics/v1.py:57: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Epoch: 01 | Epoch Time: 0m 6s
	Train Loss: 0.810 | Train f1: 0.42 | Train precision: 0.57 | Train recall: 0.34
	 Val. Loss: 0.469 |  Val. f1: 0.66 |  Val. precision: 0.80 | Val. recall: 0.57
Epoch: 02 | Epoch Time: 0m 6s
	Train Loss: 0.467 | Train f1: 0.66 | Train precision: 0.74 | Train recall: 0.61
	 Val. Loss: 0.362 |  Val. f1: 0.73 |  Val. precision: 0.81 | Val. recall: 0.68
Epoch: 03 | Epoch Time: 0m 6s
	Train Loss: 0.333 | Train f1: 0.76 | Train precision: 0.80 | Train recall: 0.72
	 Val. Loss: 0.342 |  Val. f1: 0.76 |  Val. precision: 0.80 | Val. recall: 0.72
Epoch: 04 | Epoch Time: 0m 6s
	Train Loss: 0.256 | Train f1: 0.81 | Train precision: 0.84 | Train recall: 0.79
	 Val. Loss: 0.340 |  Val. f1: 0.77 |  Val. precision: 0.82 | Val. recall: 0.73
Epoch: 05 | Epoch Time: 0m 6s
	Train Loss: 0.201 | Train f1: 0.85 | Train precision: 0.87 | Train recall: 0.84
	 Val. Loss: 0.350 |  Val. f1: 0.78 |  Val. precision: 0.82 | Val. recall: 0.75
Epoch: 06 | Epoch Time: 0m 6s
	Train Loss: 0.

In [ ]:
# Cargar el mejor modelo entrenado
model.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [ ]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

#### Evaluación

In [ ]:
prueba_loss, prueba_precision, prueba_recall, prueba_f1 = evaluate(
    model, dataloader_test, model_criterion)

print(
    f'Val. Loss: {prueba_loss:.3f} |  Val. f1: {prueba_f1:.2f} | Val. precision: {prueba_precision:.2f} | Val. recall: {prueba_recall:.2f}'
)

Val. Loss: 0.356 |  Val. f1: 0.76 | Val. precision: 0.81 | Val. recall: 0.72


#### Análisis

Las modificaciones realizadas en el modelo experimental, que incluyen la utilización de embeddings pre-entrenados, la capa GRU bidireccional y la adición de dropout, han mejorado significativamente el rendimiento en comparación con el modelo baseline. Las métricas muestran que la pérdida disminuyó de 0.393 a 0.356, indicando un mejor ajuste a los datos de test. El F1-score aumentó de 0.73 a 0.76, reflejando un mejor equilibrio entre precisión y recall. La precisión mejoró notablemente de 0.76 a 0.81, lo que sugiere que el modelo GRU tiene una mayor proporción de predicciones correctas entre todas las predicciones hechas. Aunque el recall mostró una mejora marginal de 0.71 a 0.72, este aumento indica que el modelo es un poco más eficaz en la identificación de todas las entidades relevantes en los datos. En general, estos cambios han permitido capturar mejor las relaciones contextuales en el texto, mejorando tanto la precisión como la capacidad de generalización del modelo.

### **Experimento 2**

#### Explicación

Cambios con respecto al modelo baseline:

- Utilización de embeddings pre-entrenados de "Chilean Waiting List".
- Aumento del número de capas LSTM a dos.
- Introducción de una capa de dropout con una tasa del 0.5.
- Incorporación de bidireccionalidad a las capas LSTM.


 En este experimento, se integraron embeddings pre-entrenados de "Chilean Waiting List", lo que no solo enriquece la calidad de las representaciones de las palabras, sino que también acelera el proceso de entrenamiento al proporcionar un punto de partida sólido. Además, se incrementó la complejidad del modelo al aumentar el número de capas LSTM a dos, lo que le otorga una mayor capacidad para capturar relaciones más complejas en los datos de entrada. La introducción de una capa de dropout con una tasa del 0.5 contribuyó a regularizar el proceso de entrenamiento, ayudando a controlar el sobreajuste y mejorando la capacidad de generalización del modelo. Por último, la adición de bidireccionalidad a las capas LSTM permitió que el modelo capture información contextual tanto del pasado como del futuro en la secuencia de texto, enriqueciendo así su capacidad para reconocer y clasificar entidades nombradas con mayor precisión y eficacia.

#### Implementación

In [ ]:
class NER_RNN_PRETRAINED(nn.Module):
      """
      Modelo de Recurrent Neural Netowork (RNN) para Name Entity Recognition (NER)
      utilizando embeddings pre-entrenados.

      Parameters
      ----------
      pretrained_embeddings : Tensor
            Tensor con los embeddings pre-entrenados.
      embedding_dim : int
            Dimensión de los embeddings.
      hidden_dim : int
            Dimensión de las capas ocultas de la LSTM.
      output_dim : int
            Dimensión de la capa de salida.
      n_layers : int
            Número de capas en la LSTM.
      bidirectional : bool
            Si la LSTM es bidireccional o no.
      dropout : float
            Tasa de dropout.
      pad_idx : int
            Índice de padding en los embeddings.
      """
      def __init__(self,
                  pretrained_embeddings,
                  embedding_dim,
                  hidden_dim,
                  output_dim,
                  n_layers,
                  bidirectional,
                  dropout,
                  pad_idx
                  ):

          super().__init__()

          # Capa de embedding usando embeddings pre-entrenados
          self.embedding = nn.Embedding.from_pretrained(
                                                        pretrained_embeddings,
                                                        freeze=False,
                                                        padding_idx=pad_idx
                                                        )

          # Capa LSTM
          self.lstm = nn.LSTM(
                              embedding_dim,
                              hidden_dim,
                              num_layers=n_layers,
                              bidirectional=bidirectional,
                              dropout = dropout if n_layers > 1 else 0
                              )

          # Capa de salida
          self.fc = nn.Linear(
                              hidden_dim * 2 if bidirectional else hidden_dim,
                              output_dim
                              )

          # Capa de Dropout
          self.dropout = nn.Dropout(dropout)

      def forward(self, text):
          """
          Define el paso hacia adelante de la red.

          Parameters
          ----------
          text : Tensor
              Tensor de entrada de tamaño [sent len, batch size].

          Returns
          -------
          Tensor :
              Predicciones de la red de tamaño [sent len, batch size, output dim].
          """
          #text = [sent len, batch size]

          # Aplicar embeddings y dropout
          embedded = self.dropout(self.embedding(text)) # embedded = [sent len, batch size, emb dim]

          # Pasar los embeddings por la LSTM
          outputs, (hidden, cell) = self.lstm(embedded)
          # outputs = [sent len, batch size, hid dim * n directions]
          # hidden/cell = [n layers * n directions, batch size, hid dim]

          # Pasar las salidas de la LSTM por la capa de dropout y la capa lineal
          predictions = self.fc(self.dropout(outputs)) # predictions = [sent len, batch size, output dim]

          return predictions

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# Ruta al archivo de embeddings pre-entrenados
pretrained_embedding_path = '/content/drive/MyDrive/cwlce.vec'

# Cargar embeddings pre-entrenados en formato word2vec
word_vectors = KeyedVectors.load_word2vec_format(pretrained_embedding_path, binary=False)

# Obtener vocabulario
vocab = TEXT.vocab

count=[]
# Asignar embeddings pre-entrenados a los tokens del vocabulario
for token, idx in vocab.get_stoi().items():
    # Verificar si el token está en los embeddings pre-entrenados
    if token.lower() in word_vectors.key_to_index:
        count.append(idx)
print(f"Cantidad de tokens coincidentes: {len(count)}")

Cantidad de tokens coincidentes: 12947


In [ ]:
# Dimensión de los embeddings pre-entrenados
embedding_dim = word_vectors.vector_size

# Inicializar la matriz de embeddings con ceros
embedding_matrix = torch.zeros(len(vocab), embedding_dim)

# Asignar embeddings pre-entrenados a los tokens del vocabulario
for token, idx in vocab.get_stoi().items():
    # Verificar si el token está en los embeddings pre-entrenados
    if token.lower() in word_vectors.key_to_index:
        embedding_matrix[idx] = torch.tensor(np.copy(word_vectors[token.lower()]), dtype=torch.float)
    elif token == '<PAD>':
        # Si el token es '<PAD>', asignar un vector de ceros
        embedding_matrix[idx] = torch.zeros(embedding_dim)
    else:
        # Si el token no está en los embeddings pre-entrenados, asignar un vector aleatorio
        embedding_matrix[idx] = torch.randn(embedding_dim)

In [ ]:
# INPUT_DIM = len(TEXT.vocab)
PRETRAINED_EMBEDDINGS = embedding_matrix # Utilizamos la matriz de embeddings preentrenados creada anteriormente
EMBEDDING_DIM = embedding_dim # Dimensión de los embeddings
HIDDEN_DIM = 256 # Dimensión de las capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab) # Número de clases

N_LAYERS = 2 # Número de capas en la LSTM
DROPOUT = 0.5 # Tasa de dropout
BIDIRECTIONAL = True # Indicador de si la LSTM es bidireccional

# Creamos el modelo
model_2 = NER_RNN_PRETRAINED(pretrained_embeddings=PRETRAINED_EMBEDDINGS,
                            embedding_dim=EMBEDDING_DIM,
                            hidden_dim=HIDDEN_DIM,
                            output_dim=OUTPUT_DIM,
                            n_layers=N_LAYERS,
                            bidirectional=BIDIRECTIONAL,
                            dropout=DROPOUT,
                            pad_idx=PAD_IDX)

# Nombre que tendrá el modelo guardado
model_2_name = 'experimento_2'
# Número de épocas para entrenar el modelo
model_2_n_epochs = 10

In [ ]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.get_stoi()['<PAD>'] # Definir el índice de padding para las etiquetas
model_2_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX) # Definir la función de pérdida ignorando el índice de padding

In [ ]:
# Asignar el modelo, el nombre del modelo, la función de pérdida y el número de épocas de entrenamiento
model = model_2
model_name = model_2_name
model_criterion = model_2_criterion
model_n_epochs = model_2_n_epochs

# Imprimir el número de parámetros entrenables en el modelo actual
print(f'El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

El modelo actual tiene 7,674,700 parámetros entrenables.


In [ ]:
# Aplicar la inicialización de pesos al modelo
model.apply(init_weights)

NER_RNN_PRETRAINED(
  (embedding): Embedding(16496, 300, padding_idx=0)
  (lstm): LSTM(300, 256, num_layers=2, dropout=0.5, bidirectional=True)
  (fc): Linear(in_features=512, out_features=12, bias=True)
  (dropout): Dropout(p=0.5, inplace=False)
)

In [ ]:
# Optimizador
model_optimizer = optim.Adam(model.parameters()) # SGD, AdamW

In [ ]:
# Enviar el modelo y la función de pérdida a la GPU (si está disponible)
model = model.to(device)
model_criterion = model_criterion.to(device)

In [ ]:
best_valid_loss = float('inf') # Inicializar la mejor pérdida de validación como infinita

# Bucle de entrenamiento y validación para cada época
for epoch in range(model_n_epochs):

    start_time = time.time()

    # Entrenar el modelo con el conjunto de entrenamiento
    train_loss, train_precision, train_recall, train_f1 = train(
        model, dataloader_train, model_optimizer, model_criterion)

    # Evaluar el modelo en el conjunto de validación
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model, dataloader_dev, model_criterion)

    end_time = time.time()
    # Calcular el tiempo de la época (minutos y segundos)
    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Guardar el modelo si la pérdida de validación es la mejor obtenida hasta ahora
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), '{}.pt'.format(model_name))

    # Imprimir resultados de la época
    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

Epoch: 01 | Epoch Time: 0m 8s
	Train Loss: 0.775 | Train f1: 0.43 | Train precision: 0.61 | Train recall: 0.35
	 Val. Loss: 0.457 |  Val. f1: 0.69 |  Val. precision: 0.76 | Val. recall: 0.63
Epoch: 02 | Epoch Time: 0m 8s
	Train Loss: 0.414 | Train f1: 0.70 | Train precision: 0.77 | Train recall: 0.65
	 Val. Loss: 0.358 |  Val. f1: 0.75 |  Val. precision: 0.78 | Val. recall: 0.72
Epoch: 03 | Epoch Time: 0m 7s
	Train Loss: 0.276 | Train f1: 0.80 | Train precision: 0.82 | Train recall: 0.77
	 Val. Loss: 0.342 |  Val. f1: 0.76 |  Val. precision: 0.80 | Val. recall: 0.73
Epoch: 04 | Epoch Time: 0m 8s
	Train Loss: 0.201 | Train f1: 0.85 | Train precision: 0.86 | Train recall: 0.84
	 Val. Loss: 0.355 |  Val. f1: 0.78 |  Val. precision: 0.81 | Val. recall: 0.75
Epoch: 05 | Epoch Time: 0m 7s
	Train Loss: 0.150 | Train f1: 0.89 | Train precision: 0.89 | Train recall: 0.89
	 Val. Loss: 0.385 |  Val. f1: 0.77 |  Val. precision: 0.81 | Val. recall: 0.75
Epoch: 06 | Epoch Time: 0m 7s
	Train Loss: 0.

In [ ]:
# Cargar el mejor modelo entrenado
model.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [ ]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

#### Evaluación

In [ ]:
prueba_loss, prueba_precision, prueba_recall, prueba_f1 = evaluate(
    model, dataloader_test, model_criterion)

print(
    f'Test. Loss: {prueba_loss:.3f} |  Test. f1: {prueba_f1:.2f} | Test. precision: {prueba_precision:.2f} | Test. recall: {prueba_recall:.2f}'
)

Test. Loss: 0.363 |  Test. f1: 0.76 | Test. precision: 0.79 | Test. recall: 0.73


#### Análisis

El segundo experimento mostró mejoras significativas en comparación con el modelo baseline. La pérdida se redujo de 0.393 a 0.363, indicando que el modelo del experimento tiene un mejor ajuste a los datos de prueba. El F1 score aumentó de 0.73 a 0.76, reflejando una mejora en el balance entre precisión y recall, mientras que la precisión mejoró de 0.76 a 0.79, sugiriendo que el modelo del experimento hace más predicciones correctas entre todas las predicciones realizadas. Además, el recall incrementó de 0.71 a 0.73, demostrando que el modelo del experimento es más eficaz en la identificación de las entidades relevantes.

Estas mejoras se justifican por las modificaciones introducidas. Los embeddings pre-entrenados aportaron representaciones vectoriales ricas y específicas del dominio, mejorando la comprensión contextual del modelo; el aumento a dos capas LSTM permitió capturar características más complejas y profundas del texto. La introducción de una capa de dropout con una tasa del 0.5 ayudó a prevenir el sobreajuste, forzando al modelo a generalizar mejor durante el entrenamiento. Finalmente, la bidireccionalidad en las capas LSTM permitió procesar la secuencia en ambas direcciones, capturando información contextual completa y mejorando la identificación de entidades. Estas mejoras combinadas resultaron en una mejor capacidad del modelo para aprender y generalizar, reflejadas en las métricas de desempeño mejoradas.

### **Experimento 3**

#### Explicación

Cambios con respecto al modelo baseline:

- Utilización de embeddings pre-entrenados de "Chilean Waiting List".
- Aumento del tamaño de la capa LSTM de 256 a 512.
- Introducción de un mecanismo de atención a la capa LSTM.
- Utilización de 2 capas LSTM en lugar de una.
- Aumento de la tasa de dropout de 0 (sin dropout) a 0.7.
- Incorporación de bidireccionalidad a las capas LSTM.
- Aumento del número de épocas de entrenamiento de 20 a 30.
- Cambios en el learning rate y weight decay del optimizador: de 1e-3 a 9e-4 y de 0 a 1e-5, respectivamente.

La utilización de embeddings pre-entrenados de "Chilean Waiting List" proporciona al modelo representaciones más ricas y específicas del dominio, lo que facilita la comprensión contextual del texto. El aumento del tamaño de la capa LSTM de 256 a 512 permite al modelo capturar y procesar de manera más efectiva la información en cada paso de tiempo. La introducción de un mecanismo de atención a la capa LSTM mejora la capacidad del modelo para enfocarse en partes importantes de la secuencia de entrada, permitiendo una mayor flexibilidad y precisión en la predicción. Además, la utilización de dos capas LSTM en lugar de una proporciona al modelo una mayor capacidad de representación y aprendizaje de características complejas en los datos. El aumento de la tasa de dropout a 0.7 y la incorporación de bidireccionalidad a las capas LSTM ayudan a prevenir el sobreajuste del modelo y a capturar relaciones contextuales más amplias en el texto. Finalmente, el incremento en el número de épocas de entrenamiento y los ajustes en el learning rate y weight decay del optimizador permiten un entrenamiento más robusto y una convergencia más eficiente del modelo hacia óptimos locales de rendimiento.

#### Implementación

In [ ]:
class NER_RNN_Attention(nn.Module):
      """
      Modelo de Recurrent Neural Network (RNN) junto a Attention layers para Name Entity Recognition (NER)
      utilizando embeddings pre-entrenados.

      Parameters
      ----------
      pretrained_embeddings : Tensor
            Tensor con los embeddings pre-entrenados.
      embedding_dim : int
            Dimensión de los embeddings.
      hidden_dim : int
            Dimensión de las capas ocultas de la LSTM.
      output_dim : int
            Dimensión de la capa de salida.
      n_layers : int
            Número de capas en la LSTM.
      bidirectional : bool
            Si la LSTM es bidireccional o no.
      dropout : float
            Tasa de dropout.
      pad_idx : int
            Índice de padding en los embeddings.
      """
      def __init__(self,
                  pretrained_embeddings,
                  embedding_dim,
                  hidden_dim,
                  output_dim,
                  n_layers,
                  bidirectional,
                  dropout,
                  pad_idx
                  ):

          super().__init__()

          self.embedding_dim = embedding_dim

          self.hidden_dim = hidden_dim

          self.output_dim = output_dim

          # Capa de embedding usando embeddings pre-entrenados
          self.embedding = nn.Embedding.from_pretrained(
                                                        pretrained_embeddings,
                                                        freeze=False,
                                                        padding_idx=pad_idx
                                                        )

          # Capa LSTM
          self.lstm = nn.LSTM(
                              self.embedding_dim,
                              self.hidden_dim,
                              num_layers=n_layers,
                              bidirectional=bidirectional,
                              dropout=dropout if n_layers > 1 else 0
                              )

          # Capa de atención
          self.attention = nn.Linear(self.hidden_dim * 2 if bidirectional else self.hidden_dim, 1)

          # Capa de salida
          self.fc = nn.Linear(
                              self.hidden_dim * 2 if bidirectional else self.hidden_dim,
                              self.output_dim
                              )

          # Capa de Dropout
          self.dropout = nn.Dropout(dropout)

      def forward(self, text):
          """
          Define el paso hacia adelante de la red.

          Parameters
          ----------
          text : Tensor
              Tensor de entrada de tamaño [sent len, batch size].

          Returns
          -------
          Tensor :
              Predicciones de la red de tamaño [sent len, batch size, output dim].
          """

          #text = [sent len, batch size]

          # Aplicar embeddings y dropout
          embedded = self.dropout(self.embedding(text)) #embedded = [sent len, batch size, emb dim]

          # Pasar los embeddings por la LSTM
          outputs, (hidden, cell) = self.lstm(embedded)
          #output = [sent len, batch size, hid dim * n directions]
          #hidden/cell = [n layers * n directions, batch size, hid dim]

          # Calcular las atenciones
          attn_weights = self.attention(outputs).squeeze(-1) # attn_weights = [sent len, batch size]
          attn_weights = F.softmax(attn_weights, dim=0) # attn_weights = [sent len, batch size]

          # Multiplicar las atenciones por las salidas de la LSTM
          weighted_outputs = outputs * attn_weights.unsqueeze(-1) # weighted_outputs = [sent len, batch size, hid dim * n directions]

          # Pasar las salidas ponderadas por la capa de dropout y la capa lineal
          predictions = self.fc(self.dropout(weighted_outputs)) # predictions = [sent len, batch size, output dim]

          return predictions

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# Ruta al archivo de embeddings pre-entrenados
pretrained_embedding_path = '/kaggle/working/cwlce.vec'

# Cargar embeddings pre-entrenados en formato word2vec
word_vectors = KeyedVectors.load_word2vec_format(pretrained_embedding_path, binary=False)

# Obtener vocabulario
vocab = TEXT.vocab

count=[]
# Asignar embeddings pre-entrenados a los tokens del vocabulario
for token, idx in vocab.get_stoi().items():
    # Verificar si el token está en los embeddings pre-entrenados
    if token.lower() in word_vectors.key_to_index:
        count.append(idx)
print(f"Cantidad de tokens coincidentes: {len(count)}")

Cantidad de tokens coincidentes: 12947


In [ ]:
# Dimensión de los embeddings pre-entrenados
embedding_dim = word_vectors.vector_size

# Inicializar la matriz de embeddings con ceros
embedding_matrix = torch.zeros(len(vocab), embedding_dim)

# Asignar embeddings pre-entrenados a los tokens del vocabulario
for token, idx in vocab.get_stoi().items():
    # Verificar si el token está en los embeddings pre-entrenados
    if token.lower() in word_vectors.key_to_index:
        embedding_matrix[idx] = torch.tensor(np.copy(word_vectors[token.lower()]), dtype=torch.float)
    elif token == '<PAD>':
        # Si el token es '<PAD>', asignar un vector de ceros
        embedding_matrix[idx] = torch.zeros(embedding_dim)
    else:
        # Si el token no está en los embeddings pre-entrenados, asignar un vector aleatorio
        embedding_matrix[idx] = torch.randn(embedding_dim)

In [ ]:
# INPUT_DIM = len(TEXT.vocab)
PRETRAINED_EMBEDDINGS = embedding_matrix # Utilizamos la matriz de embeddings preentrenados creada anteriormente
EMBEDDING_DIM = embedding_dim # Dimensión de los embeddings
HIDDEN_DIM = 512 #256 # Dimensión de las capas LSTM
OUTPUT_DIM = len(NER_TAGS.vocab) # Número de clases

N_LAYERS = 2 # Número de capas en la LSTM
DROPOUT = 0.7 # Tasa de dropout
BIDIRECTIONAL = True # Indicador de si la LSTM es bidireccional

# Creamos nuestro modelo
model_3 = NER_RNN_Attention(pretrained_embeddings=PRETRAINED_EMBEDDINGS,
                            embedding_dim=EMBEDDING_DIM,
                            hidden_dim=HIDDEN_DIM,
                            output_dim=OUTPUT_DIM,
                            n_layers=N_LAYERS,
                            bidirectional=BIDIRECTIONAL,
                            dropout=DROPOUT,
                            pad_idx=PAD_IDX)

# Nombre que tendrá el modelo guardado
model_3_name = 'experimento_3'
# Número de épocas para entrenar el modelo
model_3_n_epochs = 30

In [ ]:
# Loss: Cross Entropy
TAG_PAD_IDX = NER_TAGS.vocab.get_stoi()['<PAD>'] # Definir el índice de padding para las etiquetas
model_3_criterion = nn.CrossEntropyLoss(ignore_index = TAG_PAD_IDX) # Definir la función de pérdida ignorando el índice de padding

In [ ]:
# Asignar el modelo, el nombre del modelo, la función de pérdida y el número de épocas de entrenamiento
model = model_3
model_name = model_3_name
model_criterion = model_3_criterion
model_n_epochs = model_3_n_epochs

# Imprimir el número de parámetros entrenables en el modelo actual
print(f'El modelo actual tiene {count_parameters(model):,} parámetros entrenables.')

El modelo actual tiene 14,595,917 parámetros entrenables.


In [ ]:
# Aplicar la inicialización de pesos al modelo
model.apply(init_weights)

NER_RNN_Attention(
  (embedding): Embedding(16496, 300, padding_idx=0)
  (lstm): LSTM(300, 512, num_layers=2, dropout=0.7, bidirectional=True)
  (attention): Linear(in_features=1024, out_features=1, bias=True)
  (fc): Linear(in_features=1024, out_features=12, bias=True)
  (dropout): Dropout(p=0.7, inplace=False)
)

In [ ]:
# Optimizador
model_optimizer = optim.Adam(model.parameters(), lr=9e-4, weight_decay=1e-5) # SGD, AdamW

In [ ]:
# Enviar el modelo y la función de pérdida a la GPU (si está disponible)
model = model.to(device)
model_criterion = model_criterion.to(device)

In [ ]:
best_valid_loss = float('inf') # Inicializar la mejor pérdida de validación como infinita

# Bucle de entrenamiento y validación para cada época
for epoch in range(model_n_epochs):

    start_time = time.time()

    # Entrenar el modelo con el conjunto de entrenamiento
    train_loss, train_precision, train_recall, train_f1 = train(
        model, dataloader_train, model_optimizer, model_criterion)

    # Evaluar el modelo en el conjunto de validación
    valid_loss, valid_precision, valid_recall, valid_f1 = evaluate(
        model, dataloader_dev, model_criterion)

    end_time = time.time()
    # Calcular el tiempo de la época (minutos y segundos)
    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    # Guardar el modelo si la pérdida de validación es la mejor obtenida hasta ahora
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), '{}.pt'.format(model_name))

    # Imprimir resultados de la época
    print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
    print(
        f'\tTrain Loss: {train_loss:.3f} | Train f1: {train_f1:.2f} | Train precision: {train_precision:.2f} | Train recall: {train_recall:.2f}'
    )
    print(
        f'\t Val. Loss: {valid_loss:.3f} |  Val. f1: {valid_f1:.2f} |  Val. precision: {valid_precision:.2f} | Val. recall: {valid_recall:.2f}'
    )

Epoch: 01 | Epoch Time: 0m 15s
	Train Loss: 1.263 | Train f1: 0.00 | Train precision: 0.00 | Train recall: 0.00
	 Val. Loss: 1.022 |  Val. f1: 0.00 |  Val. precision: 0.00 | Val. recall: 0.00
Epoch: 02 | Epoch Time: 0m 15s
	Train Loss: 0.964 | Train f1: 0.00 | Train precision: 0.01 | Train recall: 0.00
	 Val. Loss: 0.909 |  Val. f1: 0.00 |  Val. precision: 0.02 | Val. recall: 0.00
Epoch: 03 | Epoch Time: 0m 15s
	Train Loss: 0.847 | Train f1: 0.09 | Train precision: 0.25 | Train recall: 0.06
	 Val. Loss: 0.792 |  Val. f1: 0.29 |  Val. precision: 0.49 | Val. recall: 0.21
Epoch: 04 | Epoch Time: 0m 15s
	Train Loss: 0.684 | Train f1: 0.38 | Train precision: 0.54 | Train recall: 0.30
	 Val. Loss: 0.685 |  Val. f1: 0.42 |  Val. precision: 0.52 | Val. recall: 0.35
Epoch: 05 | Epoch Time: 0m 15s
	Train Loss: 0.576 | Train f1: 0.46 | Train precision: 0.58 | Train recall: 0.38
	 Val. Loss: 0.629 |  Val. f1: 0.45 |  Val. precision: 0.52 | Val. recall: 0.39
Epoch: 06 | Epoch Time: 0m 15s
	Train Lo

In [ ]:
# Cargar el mejor modelo entrenado
model.load_state_dict(torch.load('{}.pt'.format(model_name)))

<All keys matched successfully>

In [ ]:
# Limpiar ram de cuda
torch.cuda.empty_cache()

#### Evaluación

In [ ]:
prueba_loss, prueba_precision, prueba_recall, prueba_f1 = evaluate(
    model, dataloader_test, model_criterion)

print(
    f'Test. Loss: {prueba_loss:.3f} |  Test. f1: {prueba_f1:.2f} | Test. precision: {prueba_precision:.2f} | Test. recall: {prueba_recall:.2f}'
)

Test. Loss: 0.472 |  Test. f1: 0.75 | Test. precision: 0.78 | Test. recall: 0.72


#### Análisis

Al comparar únicamente el F1-score, la precisión y el recall entre el modelo experimental y el baseline, se observa que el F1-score aumentó ligeramente de 0.73 a 0.75 en el modelo experimental, lo que indica una mejora en el equilibrio entre precisión y recall. La precisión también experimentó un pequeño incremento, pasando de 0.76 a 0.78, lo que sugiere una ligera mejora en la proporción de predicciones correctas entre todas las predicciones realizadas. Sin embargo, aunque el recall aumentó marginalmente de 0.71 a 0.72, esta mejora es mínima y podría no ser lo suficientemente significativa para impactar en la identificación efectiva de todas las entidades relevantes en los datos. Estos resultados sugieren que, a pesar de algunas mejoras en las métricas, el modelo experimental aún no logra superar de manera considerable al baseline en términos de rendimiento en la tarea de reconocimiento de entidades nombradas.

### **Comparación de experimentos**

|  Modelo  | f1-score | precision | recall |
|----------|----------|-----------|--------|
| Baseline |   0.73   |   0.76    |  0.71  |
| Modelo 1 |   0.76   |   0.81    |  0.72  |
| Modelo 2 |   0.76   |   0.79    |  0.73  |
| Modelo 3 |   0.75   |   0.78    |  0.72  |


Al analizar los resultados de los cuatro modelos, podemos observar varias tendencias interesantes.

En primer lugar, los Modelos 1 y 2 muestran un rendimiento similar en términos de F1-score, con ambos alcanzando un valor de 0.76. Sin embargo, el Modelo 1 presenta una precisión ligeramente superior (0.81 frente a 0.79) mientras que el Modelo 2 tiene un recall un poco más alto (0.73 frente a 0.72), lo que sugiere que el Modelo 1 tiende a tener menos falsos positivos y el Modelo 2 tiende a capturar más de las entidades relevantes en los datos.

Por otro lado, el Modelo 3 muestra un rendimiento ligeramente inferior en comparación con los otros modelos, con un F1-score de 0.75. Aunque la precisión es comparable al Modelo 2 (0.78 frente a 0.79), el recall es uno de los más bajos entre los modelos (0.72), lo que indica que este modelo podría estar perdiendo algunas entidades relevantes en los datos.

En general, aunque todos los modelos muestran mejoras en el rendimiento en comparación con el baseline, hay diferencias sutiles en cómo cada modelo equilibra la precisión y el recall. El Modelo 1 destaca por su alta precisión, mientras que el Modelo 2 muestra un mejor equilibrio entre precisión y recall. El Modelo 3, aunque tiene una precisión similar al Modelo 2, sufre ligeramente en términos de recall, lo que sugiere que podría necesitar ajustes adicionales para mejorar la identificación de todas las entidades relevantes.